src/common/mfu_flist_copy.c

```diff
@@ -1847,18 +1864,30 @@ static int mfu_copy_file_normal(
     }
 #endif
 
-    /* if we wrote the last chunk, truncate the file */
+    /* if sparse or using O_DIRECT, we already truncated the file at the beginning, we do not need
+     * to do it again. */
+    if (copy_opts->sparse || !copy_opts->direct)
+	    return 0;
+
+    /* if we wrote the last chunk, truncate the file if the file is larger (with O_DIRECT) */
     off_t last_written = offset + length;
     off_t file_size_offt = (off_t) file_size;
     if (last_written >= file_size_offt || file_size == 0) {
         /* Use ftruncate() here rather than truncate(), because grouplock
-         * of Lustre would cause block to truncate() since the fd is different
-         * from the out_fd. */
-        if (mfu_file_ftruncate(mfu_dst_file, file_size_offt) < 0) {
-            MFU_LOG(MFU_LOG_ERR, "Failed to truncate destination file: %s (errno=%d %s)",
-                dest, errno, strerror(errno));
-            return -1;
-        }
+	 * of Lustre would cause block to truncate() since the fd is different
+	 * from the out_fd. */
+	struct stat st;
+        int status = mfu_file_lstat(dest, &st, mfu_dst_file);
+	if (status == 0 && st.st_size > file_size_offt) {
+            if (mfu_file_ftruncate(mfu_dst_file, file_size_offt) < 0) {
+	        MFU_LOG(MFU_LOG_ERR, "Failed to truncate destination file: %s (errno=%d %s)",
+			dest, errno, strerror(errno));
+		return -1;
+	    }
```

Comment:
I'm not 100% confident, so don't want to suggest it, but I think we _**might**_ already know whether we wrote past the end because of
https://github.com/hpc/mpifileutils/blob/92ca762531f8136e3008708b3539f9b377a46256/src/common/mfu_flist_copy.c#L1795-L1811

Reply:
are you suggesting to add something like:
if (remainder > 0) { need_truncate = true} and we avoid doing the stat at the end?

Reply:
Yeah, and I also see now that we DO need to truncate for sparse if we skipped writing the last chunk. Since this relies on the truncate to _**extend**_ the file size
https://github.com/hpc/mpifileutils/blob/92ca762531f8136e3008708b3539f9b377a46256/src/common/mfu_flist_copy.c#L1813-L1820

Reply:
i updated  the PR with this